torchvision/csrc/cpu/video/Video.cpp

```diff
@@ -0,0 +1,393 @@
+        videoFPS.push_back(fps);
+        videoDuration.push_back(duration);
+      } else if (header.format.type == TYPE_AUDIO) {
+        audioFPS.push_back(fps);
+        audioDuration.push_back(duration);
+      } else if (header.format.type == TYPE_CC) {
+        ccFPS.push_back(fps);
+        ccDuration.push_back(duration);
+      } else if (header.format.type == TYPE_SUBTITLE) {
+        subsFPS.push_back(fps);
+        subsDuration.push_back(duration);
+      };
+    }
+  }
+  streamFPS.insert({{"video", videoFPS}, {"audio", audioFPS}});
+  streamDuration.insert({{"video", videoDuration}, {"audio", audioDuration}});
+
+  succeeded = Video::_setCurrentStream();
+  LOG(INFO) << "\nDecoder inited with: " << succeeded << "\n";
+  if (get<1>(current_stream) != -1) {
+    LOG(INFO)
+        << "Stream index set to " << get<1>(current_stream)
+        << ". If you encounter trouble, consider switching it to automatic stream discovery. \n";
+  }
+} // video
+
+bool Video::_setCurrentStream() {
+  double ts = 0;
+  if (seekTS > 0) {
+    ts = seekTS;
+  }
+
+  _getDecoderParams(
+      ts, // video start
+      0, // headerOnly
+      get<0>(current_stream), // stream
+      long(get<1>(
+          current_stream)), // stream_id parsed from info above change to -2
+      false // read all streams
+  );
+
+  // calback and metadata defined in Video.h
+  return (decoder.init(params, std::move(callback), &metadata));
+}
+
+std::tuple<std::string, int64_t> Video::getCurrentStream() const {
+  return current_stream;
+}
+
+std::vector<double> Video::getFPS(std::string stream) const {
+  // add safety check
+  if (stream.empty()) {
+    stream = get<0>(current_stream);
+  }
+  auto stream_tpl = _parseStream(stream);
+  std::string stream_str = get<0>(stream_tpl);
+  // check if the stream exists
+  return streamFPS.at(stream_str);
+}
+
+std::vector<double> Video::getDuration(std::string stream) const {
+  // add safety check
+  if (stream.empty()) {
+    stream = get<0>(current_stream);
+  }
+  auto stream_tpl = _parseStream(stream);
+  std::string stream_str = get<0>(stream_tpl);
+  // check if the stream exists
+  return streamDuration.at(stream_str);
+}
+
+void Video::Seek(double ts, bool any_frame = false) {
+  // initialize the class variables used for seeking and retrurn
+  video_any_frame = any_frame;
+  seekTS = ts;
+  doSeek = true;
+}
+
+torch::List<torch::Tensor> Video::Next(std::string stream) {
+  
+  bool newInit = false; // avoid unnecessary decoder initializations
+  if ((!stream.empty()) && (_parseStream(stream) != current_stream)) {
+      current_stream = _parseStream(stream);
+      newInit = true;
+  }
+
+  if ((seekTS != -1) && (doSeek == true)) {
+      newInit = true;
+      doSeek = false;
+  }
+
+  if (newInit){
+    succeeded = Video::_setCurrentStream();
+    if (succeeded) {
+      newInit = false;
+    }
+  }
+
+  // if failing to decode simply return a null tensor (note, should we
+  // raise an exeption?)
```

Comment:
I think that we should have a more explicit way of telling the user that he should stop decoding. OpenCV returns a boolean flag indicating if it was read or not, maybe we should do the same here?

Reply:
added to the discussion points